control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -1103,56 +1103,84 @@ func (r *HostedControlPlaneReconciler) reconcileKubeAPIServer(ctx context.Contex
 	}); err != nil {
 		return fmt.Errorf("failed to reconcile CAPI service admin kubeconfig secret: %w", err)
 	}
 
 	localhostKubeconfigSecret := manifests.KASLocalhostKubeconfigSecret(hcp.Namespace)
 	if _, err := controllerutil.CreateOrUpdate(ctx, r, localhostKubeconfigSecret, func() error {
-		return p.ReconcileLocalhostKubeconfigSecret(localhostKubeconfigSecret, clientCertSecret, rootCA)
+		return kas.ReconcileLocalhostKubeconfigSecret(localhostKubeconfigSecret, clientCertSecret, rootCA, p.OwnerRef, p.APIServerPort)
 	}); err != nil {
 		return fmt.Errorf("failed to reconcile localhost kubeconfig secret: %w", err)
 	}
 
 	externalKubeconfigSecret := manifests.KASExternalKubeconfigSecret(hcp.Namespace, hcp.Spec.KubeConfig)
 	if _, err := controllerutil.CreateOrUpdate(ctx, r, externalKubeconfigSecret, func() error {
-		return p.ReconcileExternalKubeconfigSecret(externalKubeconfigSecret, clientCertSecret, rootCA)
+		return kas.ReconcileExternalKubeconfigSecret(externalKubeconfigSecret, clientCertSecret, rootCA, p.OwnerRef, p.ExternalURL(), p.ExternalKubeconfigKey())
 	}); err != nil {
 		return fmt.Errorf("failed to reconcile external kubeconfig secret: %w", err)
 	}
 
 	bootstrapKubeconfigSecret := manifests.KASBootstrapKubeconfigSecret(hcp.Namespace)
 	if _, err := controllerutil.CreateOrUpdate(ctx, r, bootstrapKubeconfigSecret, func() error {
-		return p.ReconcileBootstrapKubeconfigSecret(bootstrapKubeconfigSecret, bootstrapClientCertSecret, rootCA)
+		return kas.ReconcileBootstrapKubeconfigSecret(bootstrapKubeconfigSecret, bootstrapClientCertSecret, rootCA, p.OwnerRef, p.ExternalURL())
 	}); err != nil {
 		return fmt.Errorf("failed to reconcile bootstrap kubeconfig secret: %w", err)
 	}
 
 	kubeAPIServerAuditConfig := manifests.KASAuditConfig(hcp.Namespace)
 	if _, err := controllerutil.CreateOrUpdate(ctx, r, kubeAPIServerAuditConfig, func() error {
-		return p.ReconcileAuditConfig(kubeAPIServerAuditConfig)
+		return kas.ReconcileAuditConfig(kubeAPIServerAuditConfig, p.OwnerRef, p.AuditPolicyProfile())
 	}); err != nil {
 		return fmt.Errorf("failed to reconcile api server audit config: %w", err)
 	}
 
 	kubeAPIServerConfig := manifests.KASConfig(hcp.Namespace)
 	if _, err := controllerutil.CreateOrUpdate(ctx, r, kubeAPIServerConfig, func() error {
-		return p.ReconcileConfig(kubeAPIServerConfig)
+		return kas.ReconcileConfig(kubeAPIServerConfig,
+			p.OwnerRef,
+			p.ExternalIPConfig(),
+			p.ClusterNetwork(),
+			p.ServiceNetwork(),
+			p.NamedCertificates(),
+			p.APIServerPort,
+			p.TLSSecurityProfile(),
+			p.AdditionalCORSAllowedOrigins(),
+			p.InternalRegistryHostName(),
+			p.ExternalRegistryHostNames(),
+			p.DefaultNodeSelector(),
+			p.AdvertiseAddress,
+			p.ServiceAccountIssuerURL(),
+			p.CloudProvider,
+			p.CloudProviderConfig,
+			p.EtcdURL,
+			p.FeatureGates(),
+			p.ServiceNodePortRange(),
+		)
```

Comment:
This is a case where I feel like either the function is doing too much, or the inputs need to be more coarse grained. For example, passing the NetworkSpec, a CloudProvider spec, a RegistrySpec to roll up some related state. If the function really does use essentially the _entire apiserver config_, then maybe just inverting the function to receive the Params object is alright for now (at least this way there's less chance of accidentally introducing new unrelated implied state or side effects).

Does that make sense? An argument list this long seems to be symptomatic of something and not something one would intentionally design... silver lining is that this inversion reveals such issues.

Reply:
I can even see an argument for keeping it somewhat over the top like this for now as a reminder that there's something deeper to dig into here. Stands out better than a TODO... 😆

Reply:
what are the arguments in favour of this signature vs passing down the struct `p`?

Reply:
@ironcladlou @enxebre I agree, this is over the top. The intent was to narrow down the set of parameters that apply specifically to reconciling the KAS config. Nothing more, nothing less. Passing `p` down is kind of avoiding the issue, since there's a lot more there than is needed for this config specifically. I can see creating a subset of `p` just for this one function. That way we document exactly what it consumes and don't end up with a monstruosity of a function. Wdyt?

Reply:
In my original comment I suggested the possibility of decomposing into more coarse structures as a middle ground, that's fine with me